### ISSUE TITLE
[BUG] metavuln/audit takes extremely long

### ISSUE THREAD

**Issue by csvan** [npm/cli#2197]
### Current Behavior:
the `audit` step of an install can at times take extremely long, even on a relatively strong system (Macbook Pro) with a stable and fast network connection.

The following is a sample of the `--verbose` output:

```
npm timing metavuln:cache:get:security-advisory:@graphql-toolkit/prisma-loader:59Gz6d8nXROYPTRKN6vYrOa6UJdf+ROWFyFoAwnpDagy1LQXwkmckLtH1bBwMztSTQJPAD4a0VDONdHawup+yQ== Completed in 101159ms
npm timing metavuln:load:security-advisory:@graphql-toolkit/prisma-loader:TXaH5zV/7mhOOnZrtntrv4H+++gKrBh2DyFktFZhZgQe6qTx6NVwEkZ0SItY7ve4oozE8MKKAU9vIeioVnTy6w== Completed in 3697ms
npm timing metavuln:cache:put:security-advisory:@graphql-toolkit/prisma-loader:59Gz6d8nXROYPTRKN6vYrOa6UJdf+ROWFyFoAwnpDagy1LQXwkmckLtH1bBwMztSTQJPAD4a0VDONdHawup+yQ== Completed in 7ms
npm timing metavuln:calculate:security-advisory:@graphql-toolkit/prisma-loader:TXaH5zV/7mhOOnZrtntrv4H+++gKrBh2DyFktFZhZgQe6qTx6NVwEkZ0SItY7ve4oozE8MKKAU9vIeioVnTy6w== Completed in 104863ms
npm timing metavuln:cache:put:security-advisory:@graphql-codegen/cli:8TFXQTvtJIJmXUuSXPeLbVufRY0+ZdI7ArrnI3ofZ2kV5GYJnmgUmrfkC6a5a8dCbBYC6A+vFrYusNPGDLTmjg== Completed in 3708ms
npm timing metavuln:calculate:security-advisory:@graphql-codegen/cli:TXaH5zV/7mhOOnZrtntrv4H+++gKrBh2DyFktFZhZgQe6qTx6NVwEkZ0SItY7ve4oozE8MKKAU9vIeioVnTy6w== Completed in 104865ms
npm timing metavuln:packument:graphql-config Completed in 106163ms
npm timing metavuln:load:security-advisory:graphql-config:TXaH5zV/7mhOOnZrtntrv4H+++gKrBh2DyFktFZhZgQe6qTx6NVwEkZ0SItY7ve4oozE8MKKAU9vIeioVnTy6w== Completed in 304ms
npm timing metavuln:cache:put:security-advisory:graphql-config:/YXWDVGXvr56R02NdEA9k2rayaWaPscH78yliUTlwWmTLzjskPt8ULKFDSfg11kmeGp7qiPf+AHjOaNRLR6aKg== Completed in 3ms
npm timing metavuln:calculate:security-advisory:graphql-config:TXaH5zV/7mhOOnZrtntrv4H+++gKrBh2DyFktFZhZgQe6qTx6NVwEkZ0SItY7ve4oozE8MKKAU9vIeioVnTy6w== Completed in 106470ms
npm timing metavuln:cache:get:security-advisory:@graphql-codegen/cli:0MBAUA8aqp7wX8VcRVVnz5+ZaoDmRsj9j++TTltrIffprA6IbG4B8eDD+Di6MqyqE3uxd6zhrYGssNeplS0pFw== Completed in 0ms
npm timing metavuln:load:security-advisory:@graphql-codegen/cli:pmewJCqeB28XOQi0y+FC/2l0AWypMCTlJlTtH+acej+oSSW2bW0qK7mRSLqZICG/GS4hTM4hC/1vvPM4cvR6VA== Completed in 97712ms
npm timing metavuln:cache:put:security-advisory:@graphql-codegen/cli:0MBAUA8aqp7wX8VcRVVnz5+ZaoDmRsj9j++TTltrIffprA6IbG4B8eDD+Di6MqyqE3uxd6zhrYGssNeplS0pFw== Completed in 6ms
npm timing metavuln:calculate:security-advisory:@graphql-codegen/cli:pmewJCqeB28XOQi0y+FC/2l0AWypMCTlJlTtH+acej+oSSW2bW0qK7mRSLqZICG/GS4hTM4hC/1vvPM4cvR6VA== Completed in 97718ms
npm timing metavuln:cache:get:security-advisory:react-breakpoints:0fa3kxtmIYSb4+oboS+50/jeAeY+syd54xmQRwpl/sj2EsI5c6Kn2efPF+atnb+BSSC54NJ69xrk0WqYOnTt1w== Completed in 1ms
npm timing metavuln:packument:react-breakpoints Completed in 600ms
npm timing metavuln:load:security-advisory:react-breakpoints:wFqjxLBLQtGVKqD/SUtzOK+Ygb6c2Y3YsXmJNUxHChSwjyYpy04GbsDV2nTZ53cvKmOw2FIIlGnv8LjY+XhPcA== Completed in 6ms
npm timing metavuln:cache:put:security-advisory:react-breakpoints:0fa3kxtmIYSb4+oboS+50/jeAeY+syd54xmQRwpl/sj2EsI5c6Kn2efPF+atnb+BSSC54NJ69xrk0WqYOnTt1w== Completed in 3ms
npm timing metavuln:calculate:security-advisory:react-breakpoints:wFqjxLBLQtGVKqD/SUtzOK+Ygb6c2Y3YsXmJNUxHChSwjyYpy04GbsDV2nTZ53cvKmOw2FIIlGnv8LjY+XhPcA== Completed in 609ms
npm timing metavuln:cache:get:security-advisory:@graphql-codegen/visitor-plugin-common:D00vyawQmRei4HpRbXu0R6O92BIoTGSD5DjVzVY5YaZ9g4vEakewgiO1UEcsQFyp4wN1zCqAqV1pm0B2MxVPzg== Completed in 1ms
npm timing metavuln:packument:@graphql-codegen/visitor-plugin-common Completed in 6947ms
npm timing metavuln:load:security-advisory:@graphql-codegen/visitor-plugin-common:MN7F2LvcvcKo4dUgAoSVx7aILATRm+7lsrgKV8mylGLF9ECbKK3y9fgfWwzZ6sqfTovKlRKo7vim1NoKYsyV/g== Completed in 80047ms
npm timing metavuln:cache:put:security-advisory:@graphql-codegen/visitor-plugin-common:D00vyawQmRei4HpRbXu0R6O92BIoTGSD5DjVzVY5YaZ9g4vEakewgiO1UEcsQFyp4wN1zCqAqV1pm0B2MxVPzg== Completed in 5ms
npm timing metavuln:calculate:security-advisory:@graphql-codegen/visitor-plugin-common:MN7F2LvcvcKo4dUgAoSVx7aILATRm+7lsrgKV8mylGLF9ECbKK3y9fgfWwzZ6sqfTovKlRKo7vim1NoKYsyV/g== Completed in 86999ms
npm timing metavuln:cache:get:security-advisory:relay-compiler:5agmJ2GWxgQXUqK9+MfKtokCCgfJE+aAk8HesBOUBd1dtd5qI5ugVR93N3a1wpOVttpV/wmN5EpdeRXBSqUQfQ== Completed in 0ms
npm timing metavuln:load:security-advisory:relay-compiler:yDVtim5gN8WtHkuY3Y0rFCQmuj/W1BnHk2mAoPQua8kK3xMK//7d8JxyjQAywfKjjp51DOVpO7HgZfjk7wb+1A== Completed in 19148ms
npm timing metavuln:cache:put:security-advisory:relay-compiler:5agmJ2GWxgQXUqK9+MfKtokCCgfJE+aAk8HesBOUBd1dtd5qI5ugVR93N3a1wpOVttpV/wmN5EpdeRXBSqUQfQ== Completed in 3ms
npm timing metavuln:calculate:security-advisory:relay-compiler:yDVtim5gN8WtHkuY3Y0rFCQmuj/W1BnHk2mAoPQua8kK3xMK//7d8JxyjQAywfKjjp51DOVpO7HgZfjk7wb+1A== Completed in 19151ms
npm timing metavuln:cache:get:security-advisory:@graphql-codegen/cli:ck836x3EH7RRdYlFIvb46xjwFjvbvXSidC8hiqbhkAdYe6BLlkBlgahK4m88Qgf4ddIJyPKjHeXhQ6JdU79swQ== Completed in 0ms
npm timing metavuln:load:security-advisory:@graphql-codegen/cli:2udobPgV+7cCAkwLCKvFqRrsXkgewZSjeTTcSizRSYNqvy/2qdx86AYh5V+JoUCSsiq1Y+TKUgGPhkAiQeOTGQ== Completed in 74796ms
npm timing metavuln:cache:put:security-advisory:@graphql-codegen/cli:ck836x3EH7RRdYlFIvb46xjwFjvbvXSidC8hiqbhkAdYe6BLlkBlgahK4m88Qgf4ddIJyPKjHeXhQ6JdU79swQ== Completed in 8ms
npm timing metavuln:calculate:security-advisory:@graphql-codegen/cli:2udobPgV+7cCAkwLCKvFqRrsXkgewZSjeTTcSizRSYNqvy/2qdx86AYh5V+JoUCSsiq1Y+TKUgGPhkAiQeOTGQ== Completed in 74804ms
npm timing metavuln:cache:get:security-advisory:@graphql-codegen/cli:gIfkeftjhg+WpXI43vNMIaYKZIwRL/KLtnVL3ZlLFOGYiY8yPJIy3+Xzff10xOXfFpMiQ1vCFIvq76N7jzth7g== Completed in 1ms

```
### Expected Behavior:

Audit should take a reasonable time

### Steps To Reproduce:
Run `npm install --legacy--peer-deps` on the following project:

```
{
  "name": "repro",
  "license": "UNLICENSED",
  "version": "0.0.0",
  "engines": {
    "npm": ">=7",
    "node": ">=15"
  },
  "dependencies": {
    "@fortawesome/free-solid-svg-icons": "^5.10.1",
    "@fortawesome/react-fontawesome": "^0.1.4",
    "@material-ui/icons": "^4.5.1",

    "@types/d3": "^5.7.2",
    "apollo-boost": "^0.4.7",
    "d3": "^5.15.0",
    "leaflet": "^1.5.1",
    "leaflet.locatecontrol": "^0.67.0",
    "leaflet.markercluster": "^1.4.1",
    "lodash": "^4.17.15",
    "moment": "^2.24.0",
    "prop-types": "^15.7.2",
    "query-string": "^6.11.0",
    "react-apollo": "^3.1.3",
    "react-datepicker": "^2.9.6",
    "react-intl": "^3.11.0",
    "react-leaflet": "^2.4.0",
    "react-leaflet-markercluster": "^2.0.0-rc3"
  },
  "devDependencies": {
    "@babel/preset-typescript": "^7.3.3",
    "@fortawesome/fontawesome-free": "^5.10.1",
    "@fortawesome/fontawesome-svg-core": "^1.2.21",
    "@graphql-codegen/cli": "1.11.2",
    "@graphql-codegen/typescript": "1.11.2",
    "@graphql-codegen/typescript-operations": "1.11.2",
    "@graphql-codegen/typescript-react-apollo": "1.11.2",
    "@types/leaflet": "^1.5.1",
    "@types/leaflet.locatecontrol": "^0.60.7",
    "@types/react": "^16.9.17",
    "@types/react-custom-scrollbars": "^4.0.6",
    "@types/react-datepicker": "^2.9.3",
    "@types/react-dom": "^16.8.4",
    "@types/react-leaflet": "^2.4.0",
    "@types/react-redux": "^7.1.1",
    "@types/react-router": "^5.1.4",
    "@types/react-router-dom": "^4.3.4",
    "@types/styled-components": "^5.1.4",
    "@typescript-eslint/eslint-plugin": "^2.6.1",
    "@typescript-eslint/parser": "^2.7.0",
    "awesome-typescript-loader": "^5.2.1",
    "core-js": "^3.2.1",
    "eslint-plugin-react": "^7.16.0",
    "graphql-tools": "^4.0.5",
    "react-error-boundary": "^1.2.5",
    "source-map-loader": "^0.2.4",
    "typescript": "^3.7.2",
    "webpack-merge": "^4.2.1"
  }
}
```

**Comment by lukel97** [comment]
I'm seeing this as well, especially with a lot of the `@graphql-codegen` packages

**Comment by Macavirus** [comment]
Yes, without logs this just looks like a freeze as npmcli takes up 100% of one processor core.

With logs, it becomes obvious that actions *are* happening, they just take insanely long. I would consider this a completely breaking experience that was enough to abandon any attempt to update our projects to node 15

In addition, `npm install` becomes entirely unresponsive to `SIGINT` signal sent with `Ctrl-C`

**Comment by csvan** [comment]
I can also confirm that it monopolises one core on my system and runs it at 100%

I'm interested in what it is doing that is so intensive - It was my understanding that all the security audits are actually done server-side and that the npm client only checks the status of found packages.

**Comment by aks-** [comment]
+1

**Comment by darcyclarke** [comment]
@csvan what version of `v7` do you happen to be using? We've made some improvements but know there's more to do here. If you can try upgrading to the latest v7 (ie. `npm i -g npm@7`) & seeing if theres any improvements, that would be helpful.

**Comment by csvan** [comment]
Hi @darcyclarke , I tried upgrading to npm 7.4 but I am still seeing the same behavior in the repro project unfortunately.

**Comment by csvan** [comment]
FWIW the packages with the longest audits (up to 10 minutes in some cases) all seem to be GraphQL related,

**Comment by isaacs** [comment]
This should be somewhat addressed by npm/metavuln-calculator#2.
